src/stdio.jl

```diff
@@ -32,20 +32,41 @@ end
 const bufs = Dict{String,IOBuffer}()
 const stream_interval = 0.1
 const max_bytes = 10*1024
+# max output per code cell is 512 kb
+const max_output_per_request = 1 << 19
 
-"""Continually read from (size limited) Libuv/OS buffer into an (effectively unlimited) `IObuffer`
-to avoid problems when the Libuv/OS buffer gets full (https://github.com/JuliaLang/julia/issues/8789).
-Send data immediately when buffer contains more than `max_bytes` bytes. Otherwise, if data is available
-it will be sent every `stream_interval` seconds (see the Timers set up in watch_stdio)"""
+"""
+Continually read from (size limited) Libuv/OS buffer into an `IObuffer` to avoid problems when
+the Libuv/OS buffer gets full (https://github.com/JuliaLang/julia/issues/8789). Send data immediately
+when buffer contains more than `max_bytes` bytes. Otherwise, if data is available it will be sent every
+`stream_interval` seconds (see the Timers set up in watch_stdio). Truncate the output to `max_output_per_request`
+bytes per execution request since excessive output can bring browsers to a grinding halt.
+"""
 function watch_stream(rd::IO, name::AbstractString)
     task_local_storage(:IJulia_task, "read $name task")
     try
         buf = IOBuffer()
         bufs[name] = buf
+        counter = 0
+        curr_msg = ""
         while !eof(rd) # blocks until something is available
             nb = nb_available(rd)
+            parent_id_exists = "msg_id" in keys(execute_msg.header)
+            continuation = curr_msg == execute_msg.header["msg_id"]
+            # if this a new stream then reset the byte counter
+            if parent_id_exists && !continuation
+                curr_msg = execute_msg.header["msg_id"]
+                counter = 0
+            end
             if nb > 0
-                write(buf, read(rd, nb))
+                counter += nb
+                # if this stream as surpassed the maximum output limit then ignore future bytes
+                if parent_id_exists && continuation && counter > max_output_per_request
+                    (counter - nb <= max_output_per_request) && println(orig_STDOUT[], "Excessive output truncated.")
```

Comment:
That's pretty standard Julia syntax, Stefan even recommends it here: https://stackoverflow.com/questions/22959499/julia-can-conditional-statements-evaluate-code-on-the-same-line and I see it all over the place.

The problem is that the best place to evaluate the number of bytes written is in the primary event loop. Your solution wouldn't work because it does not actually truncate future bytes which are being streamed to us. Your suggestion of resetting the counter `execute_handler` would likely work. My opposing conditionals ensured that we only print a warning message right when we overstep `max_output_per_request`, but keeps read from the libuv/os buffer so that the bytes actually get truncated, i.e. read and discarded.

I don't write it to the notebook because I'm truncating stdout and the truncation includes my warning message too. Maybe there is a more clever way of handling this.